wdl-engine/src/eval/v1/task.rs

```diff
@@ -251,186 +409,197 @@ impl<'a> TaskEvaluator<'a> {
 
-        // TODO: support retrying the task if it fails
+            let task = self.backend.spawn(state.request.clone(), tx)?;
 
-        let mut evaluated = EvaluatedTask::new(state.execution.as_ref(), status_code)?;
+            // Await the spawned notification first
+            rx.await.expect("failed to await spawned notification");
 
-        // Update the task variable's return code
-        if version >= SupportedVersion::V1(V1::Two) {
-            let task = state.scopes[TASK_SCOPE_INDEX.0]
-                .get_mut(TASK_VAR_NAME)
-                .unwrap()
-                .as_task_mut()
-                .unwrap();
-            task.set_return_code(evaluated.status_code);
-        }
+            progress(ProgressKind::TaskExecutionStarted { id });
 
-        evaluated.outputs = {
-            evaluated.handle_exit(&state.requirements)?;
+            let result = task
+                .await
+                .expect("failed to receive response from spawned task");
 
-            for index in &nodes[current..] {
-                match &graph[*index] {
-                    TaskGraphNode::Output(decl) => {
-                        self.evaluate_output(&mut state, decl, &evaluated)?;
-                    }
-                    TaskGraphNode::Input(decl) => {
-                        self.evaluate_input(&mut state, decl, inputs)?;
-                    }
-                    TaskGraphNode::Decl(decl) => {
-                        self.evaluate_decl(&mut state, decl)?;
-                    }
-                    _ => {
-                        unreachable!("only declarations should be evaluated after the command")
-                    }
-                }
+            progress(ProgressKind::TaskExecutionCompleted {
+                id,
+                result: &result,
+            });
+
+            let status_code = result?;
+            let evaluated = EvaluatedTask::new(state.request.root(), status_code)?;
+
+            // Update the task variable
+            if version >= SupportedVersion::V1(V1::Two) {
+                let task = state.scopes[TASK_SCOPE_INDEX.0]
+                    .get_mut(TASK_VAR_NAME)
+                    .unwrap()
+                    .as_task_mut()
+                    .unwrap();
+
+                task.set_attempt(retry as i64 + 1);
+                task.set_return_code(evaluated.status_code);
             }
 
-            let mut outputs: Outputs = mem::take(&mut state.scopes[OUTPUT_SCOPE_INDEX.0]).into();
-            if let Some(section) = definition.output() {
-                let indexes: HashMap<_, _> = section
-                    .declarations()
-                    .enumerate()
-                    .map(|(i, d)| (TokenStrHash::new(d.name()), i))
-                    .collect();
-                outputs.sort_by(move |a, b| indexes[a].cmp(&indexes[b]))
+            if let Err(e) = evaluated.handle_exit(state.request.requirements()) {
+                if retry >= self.config.task.retries.unwrap_or(0) {
```

Comment:
Is the zero here a magic number? Should this refer to a CONST somewhere?

Reply:
It's magic and should have a default constant like we use for cpu/memory.